### PR TITLE
Fix new cursor position when adding tags/projects on start TE

### LINKED ISSUE #2182: Start view: '#' appears in front of the description
image 1

1. Started a TE;
2. Added a description ('Stuff' in this example),
3. Added a project and then 
4. Attempting to add a tag --> the '#' appears in front of the description for some reason, which means that instead of opening the Tags list it now offers the option to create a tag out of the description I added earlier.

When adding a tag _before_ adding a project (and after adding the description) the '#' appears after the description. So that works fine.

### PR DESCRIPTION
**Pull request by daividssilverio**
Closes #2182

This is mostly a workaround for Android because the Android EditText text is returning 0 for its current `SelectionStart` when adding Projects/Tags tokens using the buttons from bottom bar on the Start TE view.

Let me know if there's a better solution for this.

🦑 whenever

### CODE REVIEW

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs [hunk 1]
```diff
@@ -1,6 +1,9 @@
 ﻿using System;
+using System.Diagnostics.SymbolStore;
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: 👀

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs [hunk 2]
```diff
@@ -90,14 +95,33 @@ private void triggerValueChanged()
...
+            var newCursorPosition =
+                (isEditingInTokenRegion || isProbablyEdittingProjectRegion(newDescription) || isProbablyEdittingTagsRegion(newDescription))
+                    ? newDescription.Length
+                    : reportedCursorPosition;
             onTextFieldInfoChanged(TextFieldInfo.WithTextAndCursor(newDescription, newCursorPosition));
         }
 
+        private bool isProbablyEdittingProjectRegion(string newDescription)
+        {
+            return ($"{lastKnownDescription.Trim()} {QuerySymbols.ProjectsString}" == newDescription.Trim()
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: `=>` will be nicer

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs [hunk 3]
```diff
@@ -90,14 +95,33 @@ private void triggerValueChanged()
...
+                    || countTokens(newDescription, QuerySymbols.Projects) <
+                    countTokens(lastKnownDescription, QuerySymbols.Projects));
+        }
+
+        private bool isProbablyEdittingTagsRegion(string newDescription)
+        {
+            return ($"{lastKnownDescription.Trim()} {QuerySymbols.TagsString}") == newDescription.Trim()
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: `=>` will be nicer

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs [hunk 4]
```diff
@@ -90,14 +95,33 @@ private void triggerValueChanged()
...
+                   || countTokens(newDescription, QuerySymbols.Tags) !=
+                   countTokens(lastKnownDescription, QuerySymbols.Tags);
+        }
+
+        private int countTokens(string text, char token)
+        {
+            return text.Count(c => c == token);
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: `=>` will be nicer

File: Toggl.Foundation.MvvmCross/Extensions/StringExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Linq;
+
+namespace Toggl.Foundation.MvvmCross.Extensions
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Extensions/StringExtensions.cs]: This is completely unrelated to Foundation, let alone MvvmCross... Move it to `Multivac`.

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs [hunk 6]
```diff
@@ -90,14 +93,31 @@ private void triggerValueChanged()
...
-            var newCursorPosition = isEditingInTokenRegion ? newDescription.Length : reportedCursorPosition;
-
+            var newCursorPosition =
+                (isEditingInTokenRegion || isProbablyEdittingProjectRegion(newDescription) || isProbablyEdittingTagsRegion(newDescription))
+                    ? newDescription.Length
+                    : reportedCursorPosition;
             onTextFieldInfoChanged(TextFieldInfo.WithTextAndCursor(newDescription, newCursorPosition));
         }
 
+        private bool isProbablyEdittingProjectRegion(string newDescription)
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: `Editing` is spelled with one `t`. In general the consonant duplication happens only if the last syllable is stressed, as for example in `adMit` (admitting). In `Edit`, the stress is on `ed`, so it's simply `editing`. English lesson complete.

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs [hunk 7]
```diff
@@ -90,14 +93,31 @@ private void triggerValueChanged()
...
+            => isNewDescriptionAddingToken(newDescription, QuerySymbols.ProjectsString)
+               || isNumberOfProjectTokensSmallerThanBefore(newDescription);
+
+        private bool isNumberOfProjectTokensSmallerThanBefore(string newDescription)
+            => newDescription.CountOccurrences(QuerySymbols.Projects) < lastKnownDescription.CountOccurrences(QuerySymbols.Projects);
+
+        private bool isProbablyEdittingTagsRegion(string newDescription)
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: I am not completely following the _probability_ logic here... 🤔 Why are these methods called `probably` something?

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs (hunk 6 above)
Comment (daividssilverio) [reply Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: 🙈 :c

File: Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs (hunk 7 above)
Comment (daividssilverio) [reply Toggl.Foundation.MvvmCross/Autocomplete/BaseTextFieldInfoTargetBinding.cs]: The user might be just adding a "@" or "#" to the TE description, that's what I was thinking.

File: Toggl.Foundation.MvvmCross/Extensions/StringExtensions.cs (hunk 5 above)
Comment (daividssilverio) [reply Toggl.Foundation.MvvmCross/Extensions/StringExtensions.cs]: Done.

**Comment by simonrozsival** [comment]
SInce all the changes are now in Android specific project (the extension method doesn't count..) I don't really have to review this anymore and you don't have to wait for my review when the others approve this.

File: Toggl.Giskard/Bindings/EditTextTextFieldInfoTargetBinding.cs [hunk 8]
```diff
@@ -51,5 +60,29 @@ protected override void MarkViewForRedrawing()
         {
             Target.Invalidate();
         }
+
+        private bool isProbablyEditingTokenRegion(string newDescription)
+            => isProbablyEditingProjectRegion(newDescription)
+               || isProbablyEditingTagsRegion(newDescription);
+
```
Comment (kornelijepetak) [inline Toggl.Giskard/Bindings/EditTextTextFieldInfoTargetBinding.cs]: multiple empty lines here 💄

Comment (daividssilverio) [reply Toggl.Giskard/Bindings/EditTextTextFieldInfoTargetBinding.cs]: Autoformat does that for me on Rider when it goes past 80 columns. How long can our lines be?

**Comment by daividssilverio** [comment]
Just to be clear guys, I'm not happy with this too, but it does fix it. This is a workaround.
I agree with @willsb about considering a rewrite of this whole logic when we finally decide what to do with #2269.
@kornelijepetak

**Pull request merged**